src/test/scala/spray/json/SnakeCaseJsonSupportSpec.scala

```diff
@@ -0,0 +1,113 @@
+package spray.json
+
+import java.util.Locale
+
+import org.specs2.mutable._
+
+class SnakeCaseJsonSupportSpec extends Specification with SnakeCaseJsonSupport {
+
+  private val PASS1 = """([A-Z]+)([A-Z][a-z])""".r
+  private val PASS2 = """([a-z\d])([A-Z])""".r
+  private val REPLACEMENT = "$1_$2"
+
+  def snakify(name: String) =
+    PASS2.replaceAllIn(PASS1.replaceAllIn(name, REPLACEMENT), REPLACEMENT).toLowerCase(Locale.US)
+
+  "SnakeCaseJsonSUpport" should {
+    "convert field names to snake cased fields names." in {
+      // Given
+      case class SampleFields(
+          HelloWorld: Int = 0,
+          A1AAbBBBBBbCCcc: Int = 0,
+          AbAAbbAAAbbb1AA2bb1AA2AAAAbb1bAA: Int = 0,
+          x: Int = 0,
+          Y: Int = 0,
+          `^There-arE_Non_(Alphabetic))_character!S`: Int = 0,
+          `tHere-Are_(AlsO)_noN_(alphaBetic))_CharaCter!s`: Int = 0,
+          `  There are space character `: Int = 0,
+          ` これは　マルチバイトAbcのaBc確認です。`: Int = 0,
+          aa: Int = 0,
+          AA: Int = 0,
+          Aa: Int = 0,
+          aA: Int = 0,
+          AAA: Int = 0,
+          AAa: Int = 0,
+          AaA: Int = 0,
+          aAA: Int = 0,
+          aaA: Int = 0,
+          aAa: Int = 0,
+          Aaa: Int = 0,
+          aaa: Int = 0
+      )
+      case class SampleFields2(
+          aaa123bbb: Int = 0,
+          aaA123bbb: Int = 0,
+          aaa123Bbb: Int = 0,
+          aaA123Bbb: Int = 0,
+          aaa123: Int = 0,
+          aaA123: Int = 0,
+          aAA123: Int = 0
+      )
+      implicit val jsonFormat1: RootJsonFormat[SampleFields] =
+         jsonFormat21(SampleFields.apply)
+      implicit val jsonFormat2: RootJsonFormat[SampleFields2] =
+        jsonFormat7(SampleFields2.apply)
+      // When
+      val sampleFields1 = SampleFields().toJson.asJsObject
+      val sampleFields2 = SampleFields2().toJson.asJsObject
+      // Then
+      val fields1 = sampleFields1.fields
+      fields1.contains(snakify("HelloWorld")) mustEqual true
```

Comment:
Good that `snakify` is a different implementation here in the test, but I would prefer if the test case expectations would be spelled out.